### ISSUE TITLE
Feature request: smarter `no-extra-non-null-assertion`

### ISSUE THREAD

**Issue by lionel-rowe** [denoland/deno_lint#1144]
Example code that should pass:

```ts
class X {
	constructor() {
		if (Math.random() > 0.5) throw new RangeError('sorry, too random')
	}

	cleanup() {
		console.log('resources have been freed')
	}
}

let x: X

try {
	x = new X()

	// ... do some stuff with x ...
} catch(e) {
	// handle errors
} finally {
	x!?.cleanup()
	// ^^^^^^^^^^^^ this line
}
```

Deno lint complains:
> Extra non-null assertion is forbidden
> Remove the extra non-null assertion operator (`!`) `deno-lint(no-extra-non-null-assertion)`

However, the `!?` is intentional: Removing the `!` causes TypeScript to complain "Variable 'x' is used before being assigned", while removing the `?` causes runtime errors if `x` doesn't exist due to erroring in the constructor.

**Comment by magurotuna** [comment]
You can work around it by putting `!` in variable declaration place, like this:

```diff
class X {
	constructor() {
		if (Math.random() > 0.5) throw new RangeError('sorry, too random')
	}

	cleanup() {
		console.log('resources have been freed')
	}
}

- let x: X
+ let x!: X
+      ^ notice ! operator here

try {
	x = new X()

	// do some stuff with x
} catch(e) {
	// handle errors
} finally {
-	x!?.cleanup()
- 	// ^^^^^^^^^^ this line
+	x?.cleanup()
+     // ^ ! operator is removed from here

}
```

However, I wouldn't do this - using `!` operator generally is a sign of code smell. In this snippet, using `!` in order to suppress the TypeScript error of using uninitialized value is completely valid complain since it is indeed possible that `x` is uninitialized in `finally` block. So I would intiialize `x` in its declaration site, like `let x: X | null = null;` so we don't have to rely on `!` operator that requires a lot of caution.

**Comment by magurotuna** [comment]
OK, it makes sense to use `!` operator in this situation to better handle some procedure in finally block while taking advantage of static analysis of TS. But like you said, it's so rare that I'm not really sure if we should fix the implementation given that it's probably not easy to implement. I'll take a look when I have the bandwidth.
Maybe it's very reasonable to add `// deno-lint-ignore` comment above the line to suppress deno_lint, along with explanation on why using `!?` cannot be avoided in this specific situation.
